### PR TITLE
das: Implement `DASState`

### LINKED ISSUE #427: das: Implement DASer public API
## Context
Currently, our `daser` only exposes Start/Stop.
 
In order to get more detailed information from the `daser`, `DASState` must be implemented.

Ideas for API methods for the `daser`:

* status of the daser(is it active/not active) 
* error handling (atm of writing, we are only logging error messages) 
*  height/hash info
* square width info

Some of the above can be modelled after [`SyncState`](#397).

## TODO

- [ ] Implement `DASState`
- [ ] Implement above suggested methods on `daser`

### PR DESCRIPTION
**Pull request by renaynay**
Resolves #427 and https://github.com/celestiaorg/celestia-node/issues/869

TODO: 

- [x] Implement square width tracking of last sampled height
- [x] IsRunning method

---------------------------------------------------------------------------------------------------------------------------

### GET `/daser`

Request

```
curl -X GET http://<ip>:26658/daser/state
```

Response

```
{"sample_routine":{"error":null,"is_running":true,"latest_sampled_height":104143,"latest_sampled_square_width":16},"catch_up_routine":{"start":"2022-07-07T15:01:25.747118+02:00","end":"0001-01-01T00:00:00Z","error":null,"id":1,"height":96247,"from":96162,"to":104142}}
```

### CODE REVIEW

File: cmd/flags_core.go [hunk 1]
```diff
@@ -73,7 +73,7 @@ func validateAddress(address string) (string, string, error) {
 		return "", "", fmt.Errorf("both protocol and host must present in the address")
 	}
 
-	if _, port, err := net.SplitHostPort(u.Host); err != nil || port == "" {
+	if _, _, err := net.SplitHostPort(u.Host); err != nil {
```
Comment (renaynay) [inline cmd/flags_core.go:76]: This check to ensure port is not empty is unnecessary, it will fail if port is empty.

File: das/state.go [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+	LatestSampledHeight() uint64
+}
+
+// state collects information about the DASer process.
+type state struct {
+	// tracks whether sampling routine is running
+	isRunning uint64
+	// indicates whether the DASer catchUp routine
+	// has successfully caught up to network head.
+	caughtUp uint64
```
Comment (adlerjohn) [inline das/state.go:18]: This is used as a Boolean, so should be prefixed with `is` to make is clear from the name.
```suggestion
	isCaughtUp uint64
```

File: das/state.go [hunk 3]
```diff
@@ -0,0 +1,54 @@
+package das
+
+import "sync/atomic"
+
+// State is an interface that allows other components to access the
+// current state of the DASer safely.
+type State interface {
```
Comment (adlerjohn) [inline das/state.go:7]: Should `IsRunning()` be added to this interface?

File: das/daser.go [hunk 4]
```diff
@@ -201,6 +214,9 @@ func (d *DASer) catchUpManager(ctx context.Context, checkpoint int64) {
 // catchUp starts a sampling routine for headers starting at the next header
 // after the `from` height and exits the loop once `to` is reached. (from:to]
 func (d *DASer) catchUp(ctx context.Context, job *catchUpJob) (int64, error) {
+	// indicate that DASer is catching up to local chain head
+	atomic.StoreUint64(&d.state.caughtUp, 0)
```
Comment (adlerjohn) [inline das/daser.go:218]: I don't really feel comfortable with dealing with literals here. How to improve:

1. Use constants instead of literals.
2. Even better, hide the setting/getting behind an interface that uses the appropriate types (i.e. `bool` in this case).

File: das/state.go (hunk 3 above)
Comment (renaynay) [reply das/state.go]: The interface was actually unused. I removed it.

File: das/state.go (hunk 2 above)
Comment (renaynay) [reply das/state.go]: done

File: das/state.go [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+
+import "sync/atomic"
+
+// state collects information about the DASer process.
+type state struct {
+	// tracks whether sampling routine is running
+	isRunning uint64
+	// indicates whether the DASer catchUp routine
+	// has successfully caught up to network head.
+	isCaughtUp uint64
```
Comment (Wondertan) [inline das/state.go:11]: While you can represent each state with booleans, adding more booleans per each new state is not sustainable + it's getting harder to reason about the code when you have multiple booleans and the relations between them.

I would suggest from the beginning starting with a simple state machine which can be extended going further:
> Stopped -> Running -> CatchingUp
With these 3 states, you can explain any state of a DASer.

File: das/daser.go [hunk 6]
```diff
@@ -153,6 +160,11 @@ func (d *DASer) sample(ctx context.Context, sub header.Subscription, checkpoint
 		log.Infow("sampling successful", "height", h.Height, "hash", h.Hash(),
 			"square width", len(h.DAH.RowsRoots), "finished (s)", sampleTime.Seconds())
 
+		d.stateLk.Lock()
+		d.state.latestSampledHeight = uint64(h.Height)
+		d.state.latestSampledSquareWidth = uint64(len(h.DAH.RowsRoots))
+		d.stateLk.Unlock()
+
```
Comment (Wondertan) [inline das/daser.go:167]: Hmm, you made a method for every access of the state, besides here.

File: das/state.go (hunk 5 above)
Comment (Wondertan) [inline das/state.go:11]: While you can represent each state with booleans, adding more booleans per each new state is not sustainable + it's getting harder to reason about the code when you have multiple booleans and the relations between them.

I would suggest from the beginning starting with a simple state machine which can be extended going further:
>Stopped -> Running -> CatchingUp -> Error

With these 4 states, you can explain any state of a DASer.

File: das/daser_test.go [hunk 7]
```diff
@@ -251,6 +251,58 @@ func TestDASer_catchUp_fails(t *testing.T) {
...
+	err := daser.Start(ctx)
+	require.NoError(t, err)
+	defer func() {
+		// wait for all "future" headers to be sampled
+		for {
+			select {
+			case <-ctx.Done():
+				t.Fatal(ctx.Err())
+			default:
+				if daser.LatestSampledHeight() == expectedFinalSampleHeight {
```
Comment (Wondertan) [inline das/daser_test.go:276]: While it is still more or less fine for a test to do a busy loop(wastes CPU constantly), the real implementation should not be doing this and there should be a way to _wait_ for the desired height to be dased.

File: cmd/flags_core.go [hunk 8]
```diff
@@ -73,7 +73,7 @@ func validateAddress(address string) (string, string, error) {
 		return "", "", fmt.Errorf("both protocol and host must present in the address")
 	}
 
-	if _, port, err := net.SplitHostPort(u.Host); err != nil || port == "" {
```
Comment (Wondertan) [inline cmd/flags_core.go:76]: Honestly, this is not part of this PR, especially if you are going to squash it into one commit.

File: das/state.go (hunk 5 above)
Comment (Wondertan) [reply das/state.go:11]: I can give some examples of how to implement it, if you want

**Review by Wondertan — CHANGES_REQUESTED** [review]
We also need to keep the error in the state if any

File: das/state.go (hunk 5 above)
Comment (renaynay) [reply das/state.go:11]: @Wondertan Yes i agree that we shouldn't be adding more booleans, but there are two clear routines running here: 
* sampling over current network headers
* sampling from last sampled checkpoint -> network head. 

It would be good to distinguish the two - from a user perspective and from a debugging perspective.

i agree though that this needs to be improved somehow bc right now `running` refers to main sample loop over current headers, and catchup refers to catchup loop.

File: das/daser_test.go (hunk 7 above)
Comment (renaynay) [reply das/daser_test.go:276]: This would require a das sub similar to how there is a heightsub.

The problem is, from a user standpoint, we already have the ability to check if a height has been dased via `CacheAvailability` and the endpoint that I've been working on here #802 

If you think this is really necessary beyond testing, I'm happy to make an issue, but for now, I wouldn't see why we need to wait for DASer to get to a certain height. For example, if a user/rollup wants to see if block at height 5000 is available and daser is still catching up at height 3400 or something, they wouldn't want to sit and wait for daser to catch up to 5000, they would just call `SharesAvailable` under the hood of the endpoint provided in #802, which will check if shares available locally, and if not, will trigger a share sampling process on the header 5000.

Comment (renaynay) [reply das/daser_test.go:276]: I'll make an issue for this as a potential enhancement but it's not high priority bc it's only testing utility for now.

**Review by adlerjohn — COMMENTED** [review]
merge conflict

File: service/share/testing.go [hunk 9]
```diff
@@ -156,3 +156,16 @@ func NewBrokenAvailability() Availability {
 func (b *brokenAvailability) SharesAvailable(context.Context, *Root) error {
 	return ErrNotAvailable
 }
+
+// TestAvailability returns an Availability that always
+// returns successfully when SharesAvailable is called.
```
Comment (adlerjohn) [inline service/share/testing.go:161]: Should this comment be on `NewTestAvailability`?

File: das/state.go [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+
+	catchUpStateLk sync.RWMutex
+	catchUpState   RoutineState // tracks information related to the `catchUp` routine
+}
+
+// RoutineState contains important information about the state of a
+// current sampling routine.
+type RoutineState struct {
+	// tracks whether routine is running
+	IsRunning uint64 `json:"is_running"`
```
Comment (vgonkivs) [inline das/state.go:26]: You can use https://github.com/uber-go/atomic/blob/master/bool.go instead of uint64

File: service/share/testing.go [hunk 11]
```diff
@@ -242,3 +242,16 @@ func NewBrokenAvailability() Availability {
 func (b *brokenAvailability) SharesAvailable(context.Context, *Root) error {
 	return ErrNotAvailable
 }
+
+type TestAvailability struct {
```
Comment (vgonkivs) [inline service/share/testing.go:246]: We already have `brokenAvailability` that always fails on `SharesAvailable`. May be we should give better naming to this one to be consistent ?

File: service/rpc/das.go [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+	LatestSampledHeight uint64 `json:"latest_sampled_height"`
+	LatestSampledWidth  uint64 `json:"latest_sampled_square_width"`
+
+	Err map[string]das.SampleError `json:"errors"`
+}
+
+func (h *Handler) handleDASStateRequest(w http.ResponseWriter, r *http.Request) {
+	dasState := new(DasStateResponse)
+	sampleState := h.das.SampleRoutineState()
+	catchUpState := h.das.CatchUpRoutineState()
```
Comment (vgonkivs) [inline service/rpc/das.go:29]: What if das state will change between fetching the state and returning the response?

File: service/rpc/das.go [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+
+import (
+	"encoding/json"
+	"net/http"
+
+	"github.com/celestiaorg/celestia-node/das"
+)
+
+const (
+	dasStateEndpoint = "/daser"
```
Comment (vgonkivs) [inline service/rpc/das.go:11]: `/daser/state` ?

Comment (renaynay) [reply service/rpc/das.go:11]: It would actually be nice to consolidate all of the current endpoints we have  / and also consolidate endpoints once we merge #802 . 

I will make an issue for this.

Comment (renaynay) [reply service/rpc/das.go:11]: Actually, the problem is, we have other endpoints like `/data_available`, `/namespaced_shares`, so I think it should be `/daser_state` maybe.

File: service/share/testing.go (hunk 11 above)
Comment (Wondertan) [reply service/share/testing.go:246]: We seem to be aligned with Dummy wording

File: service/rpc/das.go (hunk 13 above)
Comment (vgonkivs) [reply service/rpc/das.go:11]: Do we expect to have endpoints that will return other services state? Then we can have `/state/daser`, `/state/syncer`.

**Comment by Wondertan** [comment]
@renaynay, there is a race specifically in the test code

File: service/rpc/das.go (hunk 13 above)
Comment (Wondertan) [reply service/rpc/das.go:11]: +1 for daser_stare or daser/state

File: das/daser.go [hunk 14]
```diff
@@ -263,3 +284,52 @@ func (d *DASer) sampleHeader(ctx context.Context, h *header.ExtendedHeader) erro
...
+	state := d.state.sample
+	d.state.sampleLk.RUnlock()
+	return state
+}
+
+func (d *DASer) updateSampleState(h *header.ExtendedHeader, err error) {
+	height := uint64(h.Height)
+
+	d.state.sampleLk.Lock()
+	defer d.state.sampleLk.Unlock()
```
Comment (adlerjohn) [inline das/daser.go:301]: Is there a reason to use `defer` here but not in the other functions?

File: das/state.go [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+	Height uint64 `json:"height"`
+	From   uint64 `json:"from"`
+	To     uint64 `json:"to"`
+}
+
+func (ji JobInfo) Finished() bool {
+	return ji.To == ji.Height
+}
+
+func (ji JobInfo) Duration() time.Duration {
```
Comment (Wondertan) [inline das/state.go:51]: Good find. I would add this to the Syncer as well. GFI?

**Comment by renaynay** [comment]
I've extracted some modifiers (indicateRunning, indicateStopped) into methods to not crowd the otherwise long function. I can undo that for consistency, or I can keep it to maintain a cleaner looking function (`sample`).

Re one public State, I think i'll just keep it separate for now, and if there is a reason for it later, we can provide a method that returns both at once.

WDYT @Wondertan ?

Also pushing the duration method to syncer rn

**Review by vgonkivs — APPROVED** [review]

**Comment by Wondertan** [comment]
>  can undo that for consistency, or I can keep it to maintain a cleaner looking function (sample).

I am fine with the way it is right now. Consistency is important, but this is not the case where it is really important, e.g. by the following consistency in one place, you can brake it for another one.

> Re one public State, I think i'll just keep it separate for now, and if there is a reason for it later, we can provide a method that returns both at once.

It's the other way around, though. There is no reason to keep em separated, but it's fine

**Comment by Wondertan** [comment]
However, I would also fix the defer usage as per Johns's comment. Either to use defer or not to use it consistently

File: das/daser.go (hunk 14 above)
Comment (renaynay) [reply das/daser.go:301]: https://github.com/celestiaorg/celestia-node/pull/848/commits/821d5108f178fa8e84d98cf348514cbdc01b76fa done

**Review by adlerjohn — APPROVED** [review]
